### PR TITLE
Sema: optimizing compare comptime float with int

### PR DESCRIPTION
**Pull request by kkHAIKE**
relate https://github.com/ziglang/zig/pull/12901

- [x]  add case for times with NaN/Inf

### CODE REVIEW

File: test/behavior/bugs/12891.zig [hunk 1]
```diff
@@ -18,3 +18,41 @@ test "inf" {
...
+    const zero_times_inf = 0 * comptime std.math.inf(f64);
+    try std.testing.expect(std.math.isNan(zero_times_inf));
+}
+test "runtime isNan(nan * 0)" {
+    const nan_times_zero = std.math.nan(f64) * 0;
+    try std.testing.expect(std.math.isNan(nan_times_zero));
+    const zero_times_nan = 0 * std.math.nan(f64);
+    try std.testing.expect(std.math.isNan(zero_times_nan));
+}
+test " runtime isNan(inf * 0)" {
```
Comment (matu3ba) [inline test/behavior/bugs/12891.zig:53]: nit formatting.

File: test/behavior/bugs/12891.zig [hunk 2]
```diff
@@ -18,3 +18,41 @@ test "inf" {
...
+    var i: usize = 0;
+    try std.testing.expect(f < i);
+}
+test "inf >= 1" {
+    const f = comptime std.math.inf(f64);
+    var i: usize = 1;
+    try std.testing.expect(f >= i);
+}
+test "isNan(nan * 1)" {
+    const nan_times_one = comptime std.math.nan(f64) * 1;
```
Comment (matu3ba) [inline test/behavior/bugs/12891.zig:32]: why no testblock with "runtime isNan(nan *1)" ?

**Review by matu3ba — COMMENTED** [review]
some nits.

File: src/Sema.zig [hunk 3]
```diff
@@ -27678,8 +27704,10 @@ fn cmpNumeric(
             else => return Air.Inst.Ref.bool_false,
         };
         if (lhs_val.isInf()) switch (op) {
-            .gt, .neq => return Air.Inst.Ref.bool_true,
-            .lt, .lte, .eq, .gte => return Air.Inst.Ref.bool_false,
+            .neq => return Air.Inst.Ref.bool_true,
```
Comment (topolarity) [inline src/Sema.zig:27707]: I think this is incorrect for `inf == inf` and `-inf == -inf` (both should be true)

**Review by topolarity — CHANGES_REQUESTED** [review]
Thanks for the follow-up!

Just one minor change

File: src/Sema.zig (hunk 3 above)
Comment (kkHAIKE) [reply src/Sema.zig:27707]: this place is care about float vs int only...
float vs float is handled above.
we need more optimizing for it?

the test below has now passed
```zig
test "inf == inf" {
    const inf = comptime std.math.inf(f64);
    const inf2 = comptime std.math.inf(f64);
    const infr = std.math.inf(f64);
    const infr2 = std.math.inf(f64);
    const neginf = comptime -std.math.inf(f64);
    const neginf2 = comptime -std.math.inf(f64);
    const neginfr = -std.math.inf(f64);
    const neginfr2 = -std.math.inf(f64);
    try std.testing.expect(inf == inf2);
    try std.testing.expect(inf == infr);
    try std.testing.expect(infr == infr2);
    try std.testing.expect(neginf == neginf2);
    try std.testing.expect(neginf == neginfr);
    try std.testing.expect(neginfr == neginfr2);
}
```

Comment (topolarity) [reply src/Sema.zig:27707]: My mistake, I didn't realize this was mixed int/float

Sorry about that, this looks good :+1:

File: src/Sema.zig [hunk 4]
```diff
@@ -13560,9 +13560,25 @@ fn analyzeArithmetic(
...
+                        if (lhs_val.isNan()) {
+                            return sema.addConstant(resolved_type, lhs_val);
+                        }
+                        if (try lhs_val.compareWithZeroAdvanced(.eq, sema.kit(block, src))) lz: {
+                            if (maybe_rhs_val) |rhs_val| {
+                                if (rhs_val.isNan()) {
+                                    return sema.addConstant(resolved_type, rhs_val);
+                                }
+                                if (rhs_val.isInf()) {
+                                    return sema.addConstant(resolved_type, try Value.Tag.float_128.create(sema.arena, std.math.nan_f128));
```
Comment (topolarity) [inline src/Sema.zig:13577]: No need to use float_128 here - float_32 is fine

**Review by topolarity — APPROVED** [review]
Thanks again for the thorough follow-up!

I think we'll need another improvement later to fix this logic for vectors:
   1. Anywhere you use `addConstant` needs to create a `.aggregate` result if the result is supposed to be a vector
   2. `isInf/isNan` should probably become `isAllInf/isAllNan`, to support vectors
   
I'm accepting as-is though, because this is an improvement over status-quo. Vectors can be cleaned up later 🙂

**Comment by kkHAIKE** [comment]
Not sure why the test fails

**Comment by matu3ba** [comment]
@kkHAIKE You can try to rebase on top of master. master has also much more CI coverage now.

**Pull request merged**